elastic/kibana
# [HTTP] Allow default resolution to be used for certain internal routes

## Summary

When we rolled out the ability to create versioned routes the version parameters for internal routes became mandatory. However, there are several past instances where users integrating with Kibana come to rely on routes intended for internal access only, and developers may not be aware of this reliance.

In very select cases where passing version params (via headers) is not possible or cumbersome it would be useful to lift the breaking change rather than force users to, for example, abandon an upgrade.

Thus, this piece of advanced user config `useDefaultResolutionStrategyForInternalPaths` is added to enable the default version resolution to be used on select internal paths. Use like:

```yaml
server.versioned.useVersionResolutionStrategyForInternalPaths: ['/cool_path/{id?}']
```


## How to test

* Start Kibana with the following in `kibana.dev.yml`

```yaml
server.versioned:
  useVersionResolutionStrategyForInternalPaths: ['/internal/fleet/reset_preconfigured_agent_policies']
  versionResolution: 'oldest'
```

* Run this cURL command (assuming no base path mode)

```sh
curl -uelastic:changeme http://localhost:5601/internal/fleet/reset_preconfigured_agent_policies -H 'content-type: application/json' -H 'kbn-xsrf: true' -d '{}'
```

* Your request should receive `200`

Do the same but removing `useDefaultResolutionStrategyForInternalPaths` should result in `400` with an error about version parameter missing.

## Risks


* Configuration is not used as intended or is relied upon by users. The likelihood is **low** but severity **high**. To mitigate this, the configuration is designed to be cumbersome and, likely, to break over time as internal paths evolve. Additionally the configuration is disabled for serverless.


## Notes

* Added schema that will forbid this setting in serverless environments


### Checklist

Delete any items that are not applicable to this PR.
